branch-4.0: [Opt](storage) Optimize storage occupation #56815 #56846 #57220 #57327 #59075

bobhan1 opened the pull request:
### What problem does this PR solve?
pick:
- https://github.com/apache/doris/pull/56815
- https://github.com/apache/doris/pull/56846
- https://github.com/apache/doris/pull/57220
- https://github.com/apache/doris/pull/57327
- https://github.com/apache/doris/pull/59075

Clickbench result on alicloud c6.4xlarge machine, Intel(R) Xeon(R) Platinum 8269CY CPU @ 2.50GHz

||(baseline) LZ4F, 64KB data page size, 48 bucket, binary_plain_encoding_default_impl=v1, integer_type_default_use_plain_encoding =false| zstd(3), 1MB data page size, 48 bucket, binary_plain_encoding_default_impl=v1, integer_type_default_use_plain_encoding =false | zstd(3), 1MB data page size, 48 bucket, binary_plain_encoding_default_impl=v2, integer_type_default_use_plain_encoding =true |
|---|---|---|---|
|data size|16.087 GB| 11.016 GB | 8.56 GB |
|load time|592s| 555s | 539s |
|cold run total| 214.22 s | 177.57 s | 131.4 s |
|hot run total| 36.03 s | 39.82 s | 39.05 s |
### Release note

None

### Check List (For Author)

- Test <!-- At least one of them must be included. -->
    - [ ] Regression test
    - [ ] Unit Test
    - [ ] Manual test (add detailed scripts or steps below)
    - [ ] No need to test or manual test. Explain why:
        - [ ] This is a refactor/code format and no logic has been changed.
        - [ ] Previous test can cover this change.
        - [ ] No code files have been changed.
        - [ ] Other reason <!-- Add your reason?  -->

- Behavior changed:
    - [ ] No.
    - [ ] Yes. <!-- Explain the behavior change -->

- Does this need documentation?
    - [ ] No.
    - [ ] Yes. <!-- Add document PR link here. eg: https://github.com/apache/doris-website/pull/1214 -->

### Check List (For Reviewer who merge this PR)

- [ ] Confirm the release note
- [ ] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label <!-- Add branch pick label that this PR should merge into -->

bobhan1 committed:
[Opt](sys table) Add a system table column_data_sizes (#56815)

Add a system table `column_data_sizes` for querying columns' data page
sizes:
```sql
MySQL root@127.1:information_schema> show create table information_schema.column_data_sizes;
+-------------------+------------------------------------------+
| Table             | Create Table                             |
+-------------------+------------------------------------------+
| column_data_sizes | CREATE TABLE `column_data_sizes` (       |
|                   |   `BACKEND_ID` bigint NULL,              |
|                   |   `TABLE_ID` bigint NULL,                |
|                   |   `INDEX_ID` bigint NULL,                |
|                   |   `PARTITION_ID` bigint NULL,            |
|                   |   `TABLET_ID` bigint NULL,               |
|                   |   `ROWSET_ID` varchar(64) NULL,          |
|                   |   `COLUMN_UNIQUE_ID` int NULL,           |
|                   |   `COLUMN_NAME` varchar(64) NULL,        |
|                   |   `COLUMN_TYPE` varchar(64) NULL,        |
|                   |   `COMPRESSED_DATA_BYTES` bigint NULL,   |
|                   |   `UNCOMPRESSED_DATA_BYTES` bigint NULL, |
|                   |   `RAW_DATA_BYTES` bigint NULL           |
|                   | ) ENGINE=SCHEMA;                         |
+-------------------+------------------------------------------+
```
- **COMPRESSED_DATA_BYTES**: total size of compressed data pages in
bytes, for complex type(array/map/struct), it's the sum of all leaf
scalar column writers' stats
- **UNCOMPRESSED_DATA_BYTES**: total size of uncompressed data pages in
bytes, for complex type(array/map/struct), it's the sum of all leaf
scalar column writers' stats
- **RAW_DATA_BYTES**: total size of input raw data in bytes(calculated
via `XXXPageBuilder`'s `Status add(const uint8_t* vals, size_t* count)`
), for complex type(array/map/struct), it's the sum of all leaf scalar
column writers' stats

To query columns' data page sizes in a table:
```sql
SELECT
    COLUMN_NAME,
    sum(COMPRESSED_DATA_BYTES) AS compressed_data_bytes,
    sum(UNCOMPRESSED_DATA_BYTES) AS uncompressed_data_bytes,
    sum(RAW_DATA_BYTES) as raw_data_bytes
FROM information_schema.column_data_sizes
WHERE table_id=${TABLE_ID}
GROUP BY COLUMN_NAME, COLUMN_TYPE
ORDER BY data_size desc;
```

Note:
- variant column is not supported.
- In cloud mode, only BE local's rowsets will be queried.

None

- Test <!-- At least one of them must be included. -->
    - [ ] Regression test
    - [ ] Unit Test
    - [ ] Manual test (add detailed scripts or steps below)
    - [ ] No need to test or manual test. Explain why:
- [ ] This is a refactor/code format and no logic has been changed.
        - [ ] Previous test can cover this change.
        - [ ] No code files have been changed.
        - [ ] Other reason <!-- Add your reason?  -->

- Behavior changed:
    - [ ] No.
    - [ ] Yes. <!-- Explain the behavior change -->

- Does this need documentation?
    - [ ] No.
- [ ] Yes. <!-- Add document PR link here. eg:
https://github.com/apache/doris-website/pull/1214 -->

- [ ] Confirm the release note
- [ ] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label <!-- Add branch pick label that this PR
should merge into -->

bobhan1 committed:
[Opt](encoding) Add new encoding `BinaryPlainPageV2` (#56846)

1. Currently, plain binary are encoding as the following
```
|binary1|binary2|...|offset1(fixed uint32)|offset2(fixed uint32)|...
```
This PR add a new plain binary encoding which have better storage
occupation when binaries are small
```
|length1(varuint32)|binary1||length2(varuint32)|binary2|...
```
add a be conf `binary_plain_encoding_default_impl` to control.

2. move the registration of column types' encoding to `ExecEnv::_init`

bobhan1 committed:
add BinaryPlainPageV2PreDecoder by claude

fix encoding info

format

fix

update

use v2 by defaut

update

fix BinaryPlainPageV2Test

tmp fix binary dict ut

add binary dict page pre decoder and fix ut

add ut for EncodingInfo pre decoder for all types x encodings

format

reduce reduntant memcpy ini pre decoder

fix FileColumnIterator::_read_dict_data

bobhan1 committed:
[Opt](storage) Interger type use plain encoding by default (#57220)

bobhan1 committed:
use plain binary v1 by default and fuzzy storage format

hello-stephen commented:
Thank you for your contribution to Apache Doris.
Don't know what should be done next? See [How to process your PR](https://cwiki.apache.org/confluence/display/DORIS/How+to+process+your+PR).

Please clearly describe your PR:
1. What problem was fixed (it's best to include specific error reporting information). How it was fixed.
2. Which behaviors were modified. What was the previous behavior, what is it now, why was it modified, and what possible impacts might there be.
3. What features were added. Why was this function added?
4. Which code was refactored and why was this part of the code refactored?
5. Which functions were optimized and what is the difference before and after the optimization?

hello-stephen commented:
# FE UT Coverage Report
Increment line coverage `` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/57441_3ae41087237f92a8ffa2104f15ab0402428f6f04/fe_increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/57441_3ae41087237f92a8ffa2104f15ab0402428f6f04/fe_report/index.html)

bobhan1 committed:
fix beut

bobhan1 commented:
run buildall

dataroaring reviewed (dismissed):
LGTM

bobhan1 committed:
[Opt](encoding) Let storage_format control default encoding for interger and binary#59075

update

add regression

update related ut

move to tablet schema due to memtable move

remove old config

remove log

fix test

fix cloud

bobhan1 commented:
run buildall

bobhan1 committed:
fix Segment::traverse_column_meta_pbs

bobhan1 commented:
run buildall

bobhan1 commented:
run feut

bobhan1 commented:
run feut
